include/simulation_time/simulation_clock.hpp

```diff
@@ -0,0 +1,219 @@
+    //! \pre simulation_clock::set_epoch must have been called prior to calling this.
+    static std::time_t to_time_t(const simulation_time_point& sim_time);
+
+    // simulation management functions //
+
+    //! Uninitialized epoch value, aka the epoch of typename simulation_clock::calendar_clock.
+    static constexpr calendar_time_point uninitialized_epoch() noexcept;
+
+    //! Get the simulation epoch.
+    //! \see simulation_clock::calendar_time_point
+    //! \see simulation_clock::set_epoch
+    //! \returns
+    //! Calendar time point representing the simulation start date/time.
+    //! If the epoch has not been set, then simulation_clock::uninitialized_epoch
+    //! is returned.
+    static calendar_time_point epoch() noexcept;
+
+    //! Set the simulation epoch.
+    //! \note This function must only be called once.
+    //! \param calendar_time Calendar time set as the epoch.
+    //! \throws std::logic_error Thrown when this function is called more than once.
+    static void set_epoch(calendar_time_point calendar_time);
+
+    //! Check if the simulation epoch has been set.
+    //! \returns true when the epoch is set, and false otherwise.
+    static bool has_epoch() noexcept;
+
+    //! Forward the simulation by N ticks.
+    //! \param steps Number of steps to tick.
+    //! \note Ticking does *not* require that simulation_clock::set_epoch has been called.
+    //! \see simulation_clock::period
+    //! \pre `steps` must be non-negative.
+    //! \throws std::domain_error Throws when the precondition is violated.
+    static void tick(rep steps = 1);
+
+    //! Forward the simulation by some duration.
+    //! \param duration Duration to move simulation forward.
+    //! \note Ticking does *not* require that simulation_clock::set_epoch has been called.
+    //! \pre `duration` must be non-negative.
+    //! \throws std::domain_error Throws when the precondition is violated.
+    template<typename Rep, typename Period>
+    static void tick(const std::chrono::duration<Rep, Period>& duration);
+
+  private:
+    //! Simulation epoch, the calendar start time of the simulation.
+    //! \returns Mutable reference to the internal calendar epoch.
+    static calendar_time_point& internal_epoch();
+
+    //! Current simulation tick.
+    //! \returns Mutable reference to the current internal simulation tick.
+    static simulation_time_point& internal_now();
+};
+
+/* simulation_clock Implementation */
+
+inline constexpr simulation_clock::calendar_time_point simulation_clock::uninitialized_epoch() noexcept
+{
+    return {};
+}
+
+inline simulation_clock::calendar_time_point& simulation_clock::internal_epoch()
+{
+    static calendar_time_point epoch = simulation_clock::uninitialized_epoch();
+    return epoch;
+}
+
+inline simulation_clock::time_point& simulation_clock::internal_now()
+{
+    static simulation_time_point now{}; // Default initialize to zero value
+    return now;
+}
```

Comment:
Does this do the right thing with linking multiple objects that `#include <simulation_clock.hpp>`, or do we need to put this and `internal_epoch` (and presumably the rest of the implementations) in a separate `.cpp` file?

Is there some benefit to inlining all of this that's worth more than not having to think about the above question?

Reply:
Oh good catch, yeah this might be incorrect. No actual reason for it being header-only, haha